replica/table.cc

```diff
@@ -895,13 +901,14 @@ utils::chunked_vector<compaction_group*> tablet_storage_group_manager::compactio
     auto cmp = dht::token_comparator();
 
     size_t candidate_start = tr.start() ? storage_group_id_for_token(tr.start()->value()) : size_t(0);
-    size_t candidate_end = tr.end() ? storage_group_id_for_token(tr.end()->value()) : (storage_groups().size() - 1);
+    size_t candidate_end = tr.end() ? storage_group_id_for_token(tr.end()->value()) : (tablet_count() - 1);
 
     while (candidate_start <= candidate_end) {
-        auto& sg = storage_groups()[candidate_start++];
-        if (!sg) {
+        auto it = storage_groups().find(candidate_start++);
+        if (it == storage_groups().end()) {
```

Comment:
Hmm, this is pretty expensive if we had an auxiliary std::set with the occupied indices like the one you added for the tablet_sstable_set we could use to it go over only the ids in the given range.

Also, this is apparently used only for `take_storage_snapshot`, right?
And that is used only for file-based streaming of a single tablet, so the match should be of a single storage_group.  Therefore, candidate_start should be incremented by one, but rather the next one should be derived from `get_first_token` of the next tablet_id (if it's not out of range) - with that we don't need to auxiliary std::set...

Reply:
I am thinking about changing the interface to be tablet id oriented, like cleanup_tablet() so we can remove this compaction_groups_for_token_range(). file based streaming doesn't cross tablets anyway.

Maybe, for now, I can iterate through storage group instead, and use [candidate_start, candidate_end] range, to filter out. So we'd be iterating at present tablets in this shard only. It's essentially the same as finding the next tablet id, as hash map is unordered and you have to loop over elements anyway.

Reply:
But this has linear complexity.  Why not have `std::set<tablet_id>` as the one you propose for tablet_sstable_set?

Reply:
but N is a small constant, right? 100. and I suggest that we change the function to be tablet id oriented, so we can get rid of this intersection, as it's not needed in practice.

Reply:
Actually I think the impl in this version is actually good, until we change it to be tablet id oriented. if we assume that the range overlaps with 1 tablet only, then we'd be iterating at most through 2 elements (as first or last token can be exclusive).

Reply:
100 is not THAT small :)
if `[candidate_start, candidate_end]` are expected to be confined to a single tablet we should take advantage of that and look up only this tablet and nothing more.

Reply:
> Actually I think the impl in this version is actually good, until we change it to be tablet id oriented. if we assume that the range overlaps with 1 tablet only, then we'd be iterating at most through 2 elements (as first or last token can be exclusive).

Oh, what I had in mind is that candidate_start and candidate_end are tokens, not indices.
In that case this version would be fine indeed.